Enable Rustfmt

gierens opened the pull request:
The current `.rustfmt.toml` inherited from exa just disables rustfmt for no apparent reason. I know this is a big change in terms of LOCs, but having to nitpick about formatting in PRs costs a lot of time so the CI should do it automatically for us.

This re-enables rustfmt, fixes all it's issues with the current main and resulting clippy issues from fixing those. It also adds `cargo rustfmt --check` to the unit-tests workflow.

cafkafk reviewed (commented):
I mean the only reason we didn't do this IIRC was because we were scared it would impact the clap PR too much. @ariasuni do you still think we should wait on this?

cafkafk commented on src/fs/dir.rs:
```diff
@@ -39,21 +36,27 @@ impl Dir {
+    ) -> Files<'dir, 'ig> {
         Files {
-            inner:     self.contents.iter(),
-            dir:       self,
-            dotfiles:  dots.shows_dotfiles(),
-            dots:      dots.dots(),
+            inner: self.contents.iter(),
+            dir: self,
+            dotfiles: dots.shows_dotfiles(),
+            dots: dots.dots(),
```
I think that the above is preferred, but it's an unavoidable loss. Eventually we should use rustfmt for the reasons mentioned in the PR.

gierens replied:
I mean we can maybe set this in `.rustfmt.toml` so it ignores such things ... I agree I looks nicer ... but since using rustfmt I'm actually a lot less opinionated about such things as it just enforces sane defaults for formatting

cafkafk replied:
Yea same, and also for projects where I work with other people I'd rather have clear guidelines than perfect code.

gierens replied:
from quickly skimming the rustfmt configuration options, I couldn't find any way to force this indentation ... we could maybe use `#[rustmt::skip...]` but I'm not sure it's really worth it.

ariasuni replied:
Personally, I think slightly worse looking is worth more readable diffs (= easier code reviews)

cfxegbert replied:
For large tables, like filetype mapping or icon mapping, we should turn off formatting with `#[rustfmt::skip]`.

PThorpe92 replied:
Rustfmt is a must. i do agree with @cfxegbert .. `#[rustfmt:skip]` for large tables, where any edits will be carefully reviewed for formatting

gierens replied:
I'm gonna skim the code for such places and try to address those with skips.

gierens replied:
Alright, so I added skips where it seemed appropriate. Since attributes on expressions are still experimental however, there is a bit of a trade-off. There are quite a few functions evaluating to a `match` without explicit `return` statement. In those cases there are two options:
1. just let rustfmt indent it how it wants it.
2. make a statement of it with `return`:
```rust
    pub fn format(self, time: &DateTime<FixedOffset>) -> String {
        #[rustfmt::skip]
        return match self {
            Self::DefaultFormat  => default(time),
            Self::ISOFormat      => iso(time),
            Self::LongISO        => long(time),
            Self::FullISO        => full(time),
            Self::Relative       => relative(time),
        };
    }
```

I went with 1 when the indentation doesn't improve readability a lot, like in two option matches, and with 2 for larger constructs where the indentation actually helps quite a bit. The example above might have actually been more one for option 1.

Opinions?

PThorpe92 replied:
I think outside of the large tables, and `flags.rs`... it might be easy to overthink it. I think just letting RustFmt do its thing, is perfectly fine :+1:

ariasuni replied:
Well I’d say most places where there’s `#[rustfmt::skip]` could have done without it, but if you believe it really is better, than let’s go with that (anyway that’s only a fraction of the code…).

gierens replied:
Yeah, I started of applying them more happily in the beginning and later tried to avoid it if possible so not all might be necessary. At least we are not losing anything and can still remove some of them later on with no trouble.

gierens commented:
Ah, yeah I though about that too, but I think if we also just apply rustfmt in that branch we shouldn't have any issues there if I'm not mistaken.

ariasuni commented:
I’m not sure how stable are the hash in Git after a merge, so I’m not sure you can do it in this PR, but I would like the commit hash to be put in a `.git-ignore-revs-file` so that people can configure Git locally to ignore this commit when using `blame`.

cafkafk commented:
> but I feel like we’re postponing a lot of things because of the clap PR which is a pain…

Sounds like a tentative yes?

sbatial commented:
Would love to see this.
I don't know if we talked about it back then but I was surprised to see the `.rustfmt.toml` as well when going over the clippy suggestions.

---

> I’m not sure how stable are the hash in Git after a merge, so I’m not sure you can do it in this PR, but I would like the commit hash to be put in a .git-ignore-revs-file so that people can configure Git locally to ignore this commit when using blame.

TIL! git has always another card up it's sleeve

cfxegbert commented on src/output/icons.rs:
```diff
@@ -7,103 +7,103 @@ use crate::fs::File;
 struct Icons;
 
 impl Icons {
```
Can we turn if off just for this list of constants?

gierens replied:
yep, probably ... gonna skim the code for those places and try to use rustfmt's skips there.

gierens replied:
fixed

ariasuni commented:
I’m fine with this being merged, I just don’t want @MartinFillon to be discouraged to the rebase needed afterwards. If the clap PR is being merged soon, re-applying `rustfmt` on everything should not be a lot of work.

MartinFillon commented:
It is okay @ariasuni next week I will have the time to work on eza more than this week.

cafkafk reviewed (changes requested):
I noticed the flake CI picked this up, when I run `nix fmt` I get all of these changes https://gist.github.com/cafkafk/204fa53dd5018b3d053f85404e04d0b8

Seems like there are some of those that should also be ignored.

gierens commented:
> I noticed the flake CI picked this up, when I run `nix fmt` I get all of these changes https://gist.github.com/cafkafk/204fa53dd5018b3d053f85404e04d0b8
> 
> Seems like there are some of those that should also be ignored.

I just noticed I have `nix` installed on my Arch so I can reproduce this locally, great. I will get on this after rebasing on the current main.

gierens commented:
@cafkafk I rebased on main and fixed the treefmt issues ... I also had to add `src/options/flags.rs` to excludes in `treefmt.nix` as this seems to ignore the rustfmt skip in the code for that file.

cafkafk reviewed (approved):
LGTM
